sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/BeamCalciteSchemaFactory.java

```diff
@@ -17,28 +17,164 @@
  */
 package org.apache.beam.sdk.extensions.sql.impl;
 
+import com.google.common.collect.ImmutableMap;
+import java.util.Collection;
+import java.util.Collections;
 import java.util.Map;
+import java.util.Properties;
 import java.util.ServiceLoader;
+import java.util.Set;
+import org.apache.beam.sdk.extensions.sql.meta.provider.ReadOnlyTableProvider;
 import org.apache.beam.sdk.extensions.sql.meta.provider.TableProvider;
 import org.apache.beam.sdk.extensions.sql.meta.store.InMemoryMetaStore;
 import org.apache.beam.sdk.extensions.sql.meta.store.MetaStore;
+import org.apache.calcite.jdbc.CalciteConnection;
+import org.apache.calcite.linq4j.tree.Expression;
+import org.apache.calcite.rel.type.RelProtoDataType;
+import org.apache.calcite.schema.Function;
 import org.apache.calcite.schema.Schema;
 import org.apache.calcite.schema.SchemaFactory;
 import org.apache.calcite.schema.SchemaPlus;
+import org.apache.calcite.schema.SchemaVersion;
+import org.apache.calcite.schema.Table;
 
-/** Factory that creates a {@link BeamCalciteSchema}. */
-public class BeamCalciteSchemaFactory implements SchemaFactory {
-  public static final BeamCalciteSchemaFactory INSTANCE = new BeamCalciteSchemaFactory();
+/**
+ * Factory classes that Calcite uses to create initial schema for JDBC connection.
+ *
+ * <p>Calcite creates a factory using default constructor and isn't flexible enough to hook into the
+ * process to pass in additional parameters (e.g. connection), so we have to add the indirection
+ * layer.
+ *
+ * <p>{@link LoadAllProviders} is the default factory that greedily finds and loads all available
+ * table providers. This is used in a normal JDBC path, e.g. when CLI connects to {@link
+ * JdbcDriver}.
+ *
+ * <p>{@link EmptySchema} is used by {@link JdbcDriver#connect(TableProvider)} to avoid loading all
+ * available table providers.
+ */
+public class BeamCalciteSchemaFactory {
+
+  /**
+   * Called by {@link JdbcConnection} when initializing to convert the initial empty schema to
+   * actual {@link BeamCalciteSchema}.
+   */
+  static TableProvider fromInitialEmptySchema(JdbcConnection jdbcConnection) {
+    InitialEmptySchema initialEmptySchema = jdbcConnection.getCurrentBeamSchema();
+    return initialEmptySchema.toTableProvider(jdbcConnection);
+  }
+
+  /**
+   * Loads all table providers using service loader. This is the default configured in {@link
+   * JdbcDriver#connect(String, Properties)}.
+   */
+  public static class LoadAllProviders extends InitialEmptySchema implements SchemaFactory {
+
+    @Override
+    public TableProvider toTableProvider(JdbcConnection connection) {
+      MetaStore metaStore = new InMemoryMetaStore();
+      for (TableProvider provider :
```

Comment:
If external table provider also extends table provider interface and uses AutoService, then this by default loading all in-memory and external table provider to the default `beam` schema, which will violate multi schema design (and lose the benefits of having multiple schema).

So it implies that external table provider will not use table provider interface. We can address the interface design late when we will have the first external table provider.

Reply:
Good catch. Yes, if an external table provider implements the interface, then it will get loaded here, I haven't thought about it. I can think of multiple ways to handle it. One is, as you suggested, to have a separate interface, another is to introduce something like a "@DoNotAutoLoad" annotation. I think it can be addressed separately.

Reply:
Sounds good to me. 

This PR is not for external table provider interface design so we can address it later.

Reply:
Can you leave a TODO or comment here as a reminder of the issue?

Reply:
I don't think a todo here will be helpful. For example if we choose to introduce a new interface we might not even look at this line ever again and forget about the todo

Reply:
True, but if someone doesn't have/remember the context, chooses a table provider interface for their external table provider and could not have the right behavior, the person will trace down to this code piece. At leaving a comment will help understand the situation. 

Maybe just leave a comment say: "All classes implement TableProvider interface and use AutoService will be registered into default 'beam' schema".

I personally had trouble to understand how table provider was loaded into Calcite due to this less clear auto service loading way.

Reply:
clarified in the comments